### PR TITLE
Use std::filesystem::path

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

This PR replaces uses of `std::string` with `std::filesystem::path` when representing filesystem paths. This better expresses intent and opens the door for better supporting non-ASCII file paths. Because `std::filesystem::path` can freely be converted to and from strings, this change doesn't introduce much more type safety but also minimally breaks existing code that is using strings to represent paths.

This PR includes a few instances where I had to refactor some small functions that used `std::string` member functions and were easily replaced by `std::filesystem::path` member functions, in particular `std::filesystem::path::extension()`.

## Tasks

* [ ] Tested on Linux
* [ ] Tested on Windows
* [x] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
```
In file included from /home/runner/work/SFML/SFML/src/SFML/System/FileInputStream.cpp:28:
/home/runner/work/SFML/SFML/include/SFML/System/FileInputStream.hpp:37:10: fatal error: 'filesystem' file not found
#include <filesystem>
         ^~~~~~~~~~~~
```
The Android build isn't happy with using `<filesystem>`. Anyone know what this is about? @vittorioromeo Is the only one I know of to have used this header in the master branch.

**Comment by ChrisThrasher** [comment]
I'm also running into plenty of issues with `std::filesystem::path::c_str()` returning a type that isn't convertible to `const char*` which is breaking external APIs that expect `const char*` paths. Not yet sure how to handle this. I guess we'll have to just brute force convert the path to an ASCII representation and hope that we're not discarding non-ASCII characters that may be in the path.

**Comment by Eremiell** [comment]
Update of the Android toolchain may help with the issue, the latest NDK r23b has clang12 toolchain. The failed build above seems to suggest the currently used one sports gcc4.9. https://github.com/android/ndk/wiki/Changelog-r23

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Network/Ftp.cpp
Comment (reconndev) [inline src/SFML/Network/Ftp.cpp:302]: Since you're already introducing `<filesystem>` it would be better to use proper functions instead of only converting to string every time. `std::filesystem::path` is already implicitly convertible to `std::string` so only that change would not introduce any benefit at all:
```cpp
void usePath(const std::string& path);

usePath(std::filesystem::current_path());  // works
```

Here the comment says: `// Extract the filename from the file path` so maybe the intention was `remoteFile.filename()` or `remoteFile.stem()`

File: src/SFML/System/FileInputStream.cpp
Comment (reconndev) [inline src/SFML/System/FileInputStream.cpp:70]: `std::fopen` does not work on Windows? The cppreference documentation seems to state it is possible.

Comment (eXpl0it3r) [reply src/SFML/System/FileInputStream.cpp:70]: I don't know what the "right" solution is, but handling non ASCII characters is still a major issue and I think `std::fopen` can't handle it properly. Not sure if `c_str()` however defeats this. I could expect a `filename.wstring()` here if we go with `_wfopen`.

File: src/SFML/Network/Ftp.cpp
Comment (ChrisThrasher) [reply src/SFML/Network/Ftp.cpp:302]: `std::filesystem::path` is convertible to `std::string` on Linux and macOS but not on Windows. I had to add `.string()` all over the place to appease the Windows CI jobs that kept failing without that. Maybe we should using `auto` in more contexts so Windows builds can automatically use wide character strings but other OSes can use `std::string`.

File: src/SFML/System/FileInputStream.cpp
Comment (ChrisThrasher) [reply src/SFML/System/FileInputStream.cpp:70]: `std::filesystem::path::c_str()` returns a `const wchar_t*` on Windows which is not convertible to the `const char*` that `std::fopen` needs. This is one instance of a larger problem that Windows paths use wide characters and many external APIs we're using (including `std::fopen`) use `const char*` for paths. You can do a lossy conversion like `path.string().c_str()` which is guaranteed to convert to a `std::string` then a `const char*` but that may break code that actually needs to use non-ASCII characters in paths. 

Because there's apparently a Windows-specific analog to `std::fopen` I added preprocess logic to use that when appropriate but in most other circumstances, the lossy conversion is the best option I could come up with.

**Comment by ChrisThrasher** [comment]
The iOS build is complaining about some `std::filesystem` utilities not being available until iOS 13.
```
/Users/SFML/Desktop/buildbot/ios-clang-64/build/src/SFML/System/FileInputStream.cpp:72:38: error: 'c_str' is unavailable: introduced in iOS 13.0
    m_file.reset(std::fopen(filename.c_str(), "rb"));
```
Does this warrant upgrading our iOS requirement to 13? That's a few years old now.

File: src/SFML/Network/Ftp.cpp
Comment (ChrisThrasher) [reply src/SFML/Network/Ftp.cpp]: > it would be better to use proper functions instead of only converting to string every time
You're right. I was just too lazy to replace this code without whatever the `<filesystem>` counterpart is. I just updated the PR doing exactly that. It's a lot shorter and easier to read now.

File: src/SFML/System/FileInputStream.cpp
Comment (reconndev) [reply src/SFML/System/FileInputStream.cpp:70]: Right, I checked and unless `std::fstream` is used, one has to work around using macros like this, sadly.

Comment (ChrisThrasher) [reply src/SFML/System/FileInputStream.cpp:70]: I feel like once `std::filesystem` support lands in the API, there will be a whole second effort testing and validating that non-ASCII paths can be used if that's even possible. I know that Freetype and OpenAL are two of the external APIs we're using that expects `const char*` paths so there may be limitation to how widespread we can make non-ASCII support if 3rd party APIs don't support it.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
Looks like the last problem is the SFML-Buildbot iOS build job failing because it's targeting something older than iOS 13 where `<filesystem>` was added. We think the problem is that the machine running that CI job is too old to have a version of Xcode that can target iOS 13 or newer so we suspect upgrading to a newer macOS version will fix that problem.

Does anyone know anything about the machine running that job or how we might fix the issue of iOS 12 (or older) being used still?

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
I think we're good now! Thanks to help from @Eremiell we were able to work through issues caused by the Android and iOS toolchains being too out of date to work with C++17's `<filesystem>`. Now the CI pipeline is happy with these changes. If the maintainers would prefer, the earlier commits in this PR can be broken out into some number of smaller PRs but I think it's fitting to keep this all together since the updates to the toolchains are only necessitated by the addition of `<filesystem>`.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Audio/ALCheck.cpp
Comment (eXpl0it3r) [inline src/SFML/Audio/ALCheck.cpp:45]: Does this work as expected? Why didn't we use `std::string` on the function directly in the past? 🤔

File: src/SFML/Audio/AudioDevice.cpp
Comment (eXpl0it3r) [inline src/SFML/Audio/AudioDevice.cpp:111]: Is this really a path and not just the extension (e.g. .flac)?

File: include/SFML/Network/Ftp.hpp
Comment (eXpl0it3r) [inline include/SFML/Network/Ftp.hpp:509]: Are these *always* valid paths?

File: src/SFML/Graphics/GLCheck.cpp
Comment (eXpl0it3r) [inline src/SFML/Graphics/GLCheck.cpp:38]: Same question as for alCheckError

File: src/SFML/Graphics/ImageLoader.cpp
Comment (eXpl0it3r) [inline src/SFML/Graphics/ImageLoader.cpp:250]: Does it need to be a  `std::string`?

File: src/SFML/Audio/AudioDevice.cpp
Comment (ChrisThrasher) [reply src/SFML/Audio/AudioDevice.cpp:111]: I used `std::filesystem::path` because `std::filesystem::path::extension` returns a `std::filesystem::path`. Apparently that's how the standard library views extensions.

File: src/SFML/Audio/ALCheck.cpp
Comment (ChrisThrasher) [reply src/SFML/Audio/ALCheck.cpp:45]: As far as I can tell it's fine. No idea why we didn't use `std::string` before.

File: include/SFML/Network/Ftp.hpp
Comment (ChrisThrasher) [reply include/SFML/Network/Ftp.hpp:509]: What do you mean?

File: src/SFML/Graphics/GLCheck.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/GLCheck.cpp:38]: Seems fine ¯\_(ツ)_/¯

File: src/SFML/Graphics/ImageLoader.cpp
Comment (ChrisThrasher) [reply src/SFML/Graphics/ImageLoader.cpp:250]: A quick local test shows that `extension` could be a `std::filesystem::path` and the comparisons to character arrays on later lines would still compile and presumably behave the same.

File: include/SFML/Network/Ftp.hpp
Comment (eXpl0it3r) [reply include/SFML/Network/Ftp.hpp:509]: I don't know exactly, just wondering if the FTP protocol enforces normal paths or if there are also cases where you might not have "normal" paths but just some random strings 🤔

Comment (ChrisThrasher) [reply include/SFML/Network/Ftp.hpp:509]: I see. `std::filesystem::path` does normalize the path to remove unnecessary bits like `../` or multiple consecutive backslashes since those things add no extra info to a path. Maybe that removal of characters could break something here. I'm not sure.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
If this gets merged we should also be able to remove the GCC 8 workarounds in the unit tests that remove `<filesystem>` for that compiler.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

File: include/SFML/Network/Ftp.hpp
Comment (vittorioromeo) [reply include/SFML/Network/Ftp.hpp:508]: Has any research/testing been done here? Maybe worthwhile revert the changes here to merge the PR

Comment (ChrisThrasher) [reply include/SFML/Network/Ftp.hpp:508]: No I haven't tested this particular snippet. I'll revert this for now.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Review by vittorioromeo — APPROVED** [review]
LGTM, thank you so much!

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]
Looks good to me. Thank you a lot for all this effort! 🙂 

Btw. did you ever give the FTP client a test run, after these changes?

**Comment by vittorioromeo** [comment]
> Looks good to me. Thank you a lot for all this effort! 🙂
> 
> Btw. did you ever give the FTP client a test run, after these changes?

I had @ChrisThrasher remove the code that you were concerned about, but I do agree it would be a good idea to test it anyway.

**Comment by ChrisThrasher** [comment]
> Btw. did you ever give the FTP client a test run, after these changes?

I did not. I did however revert one instance of `filesystem` in the FTP class which looked suspicious.

**Comment by vittorioromeo** [comment]
I'll merge and try out the FTP example on `master`. Will let you know if there is anything fishy...

File: src/SFML/Audio/AudioDevice.cpp
Comment (ChrisThrasher) [reply src/SFML/Audio/AudioDevice.cpp:112]: I think you're totally right. Now that I look more deeply, I don't see any references to filesystem extensions. See #2024. Thanks!

File: include/SFML/Network/Ftp.hpp
Comment (ChrisThrasher) [reply include/SFML/Network/Ftp.hpp]: On second inspection, `std::filesystem::path::path` does not appear to modify the string you give it upon construction. It merely stores that string. Any manipulation of that path happens later upon request when calling functions like [`std::filesystem::canonical`](https://en.cppreference.com/w/cpp/filesystem/canonical). I believe it's safe to use `std::filesystem` here.

**Pull request merged**
